xbmc/addons/binary-addons/BinaryAddonManager.cpp

```diff
@@ -172,105 +413,68 @@ bool CBinaryAddonManager::AddAddonBaseEntry(BINARY_ADDON_LIST_ENTRY& entry)
-  m_enabledAddons.erase(base->ID());
-
-  /**
-   * @todo add way to inform type addon manager (e.g. for PVR) and parts about changed addons
-   *
-   * Currently only Screensaver and Visualization use the new way and not need informed.
-   */
+  {
+    CLog::Log(LOGERROR, "CBinaryAddonManager::%s - timed out", __FUNCTION__);
+  }
 }
 
-void CBinaryAddonManager::InstalledChangeEvent()
+void CBinaryAddonManager::OnPostInstallEvent(const std::string& addonId, bool update, bool modal)
 {
-  BINARY_ADDON_LIST binaryAddonList;
-  CAddonMgr::GetInstance().GetInstalledBinaryAddons(binaryAddonList);
-
-  CSingleLock lock(m_critSection);
+  MsgAddonPostInstall* msg = new MsgAddonPostInstall;
+  msg->addonID = addonId;
+  msg->update = update;
+  msg->modal = modal;
+  m_control.SendOutMessage(CBinaryAddonDataProtocol::POST_INSTALL_EVENT, &msg, sizeof(MsgAddonPostInstall*));
+}
 
-  BinaryAddonMgrBaseList deletedAddons = m_installedAddons;
-  for (auto addon : binaryAddonList)
+void CBinaryAddonManager::OnPreUnInstallEvent(const std::string& addonId)
+{
+  Actor::Message *reply;
+  if (m_control.SendOutMessageSync(CBinaryAddonDataProtocol::PRE_UNINSTALL_EVENT, &reply, 5000, const_cast<char*>(addonId.c_str()), addonId.size()+1))
   {
-    auto knownAddon = m_installedAddons.find(addon.second.ID());
-    if (knownAddon == m_installedAddons.end())
+    bool success = reply->signal == CBinaryAddonDataProtocol::ACC;
+    reply->Release();
+    if (!success)
     {
-      CLog::Log(LOGDEBUG, "CBinaryAddonManager::%s: Adding new binary addon '%s'", __FUNCTION__, addon.second.ID().c_str());
-
-      if (!AddAddonBaseEntry(addon))
-        continue;
-
-      /**
-       * @todo add way to inform type addon manager (e.g. for PVR) and parts about changed addons
-       *
-       * Currently only Screensaver and Visualization use the new way and not need informed.
-       */
-    }
-    else
-    {
-      deletedAddons.erase(addon.second.ID());
+      CLog::Log(LOGERROR, "CBinaryAddonManager::%s - returned error", __FUNCTION__);
     }
   }
-
-  for (auto addon : deletedAddons)
+  else
   {
-    CLog::Log(LOGDEBUG, "CBinaryAddonManager::%s: Removing binary addon '%s'", __FUNCTION__, addon.first.c_str());
-
-    m_installedAddons.erase(addon.first);
-    m_enabledAddons.erase(addon.first); // Normally should the addon disabled by another event, but to make sure also erased here
-
-    /**
-     * @todo add way to inform type addon manager (e.g. for PVR) and parts about changed addons
-     *
-     * Currently only Screensaver and Visualization use the new way and not need informed.
-     */
+    CLog::Log(LOGERROR, "CBinaryAddonManager::%s - timed out", __FUNCTION__);
   }
 }
+
+void CBinaryAddonManager::OnPostUnInstallEvent(const std::string& addonId)
+{
+  m_control.SendOutMessage(CBinaryAddonDataProtocol::POST_UNINSTALL_EVENT, const_cast<char*>(addonId.c_str()), addonId.size()+1);
```

Comment:
Like them is only non blocked for CAddonManager. But with the message can the CBinaryAddonManger do his work.